### DISCUSSION TITLE
Can't use context on the server with React PDF

### DISCUSSION THREAD

**Discussion by yousefelgoharyx** [vercel/next.js#68553] [Help]
### Link to the code that reproduces this issue

https://codesandbox.io/p/devbox/holy-rgb-xrw7h8

### To Reproduce

Click the "create" button. This button should generate a pdf on the server

### Current vs. Expected behavior

The PDF should be generated correctly but instead an error regarding client and server components is thrown because I'm using React Context
This error shouldn't be thrown since i'm not rendering anything on the client, just using React context to render a pdf

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.5.0: Wed May  1 20:12:58 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T6000
  Available memory (MB): 16384
  Available CPU cores: 10
Binaries:
  Node: 20.11.1
  npm: 10.2.4
  Yarn: 1.22.22
  pnpm: 9.6.0
Relevant Packages:
  next: 14.2.3 // There is a newer version (14.2.5) available, upgrade recommended!
  eslint-config-next: 14.2.3
  react: 18.3.1
  react-dom: 18.3.1
  typescript: 5.5.3
Next.js Config:
  output: N/A
 ⚠ There is a newer version (14.2.5) available, upgrade recommended!
   Please try the latest canary version (`npm install next@canary`) to confirm the issue still exists before creating a new issue.
   Read more - https://nextjs.org/docs/messages/opening-an-issue
```

### Which area(s) are affected? (Select all that apply)

Module Resolution, Runtime

### Which stage(s) are affected? (Select all that apply)

next dev (local)

**Comment by yousefelgoharyx** [comment]
Ok I found a solution. First let's understand the problem
As far as I understand and I could be wrong, Next.js messes with every import statement you have

1. If you don't have "use client" in a component file with "client-side" features, it immediately throws an error
2. If you add "use client" and import the component in another file that runs on the server environment and try to use it say in ReactPDF.render() it will throw since the value of the imported component gets transformed to a reference afaik

This is very understandable since server components need "only" a reference to client components but what is the solution in the case you actually need the component not a reference to it?

My workaround is to use pages directory since Next.js does not mess with any imports there

@leerob @timneutkens 
What could be the solution here?

**Comment by samcx** [comment]
@yousefelgoharyx I don't think there's an issue with React Context usage with App Router, so I have moved this to Discussions. We do have some documentation related to this → https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#using-context-providers.

Let me know if that helps!

**Reply by yousefelgoharyx** [reply]
It's definitely an issue with app router
Rendering a pdf in pages directory works just fine
What makes you say it's not an issue with app router?

**Reply by yousefelgoharyx** [reply]
To be very specific 
There's no issue using context in app router
There's an issue using context anywhere else
And I believe app router is the problem here
I'm open to discuss this further and offer ideas in mind to fix this

**Reply by samcx** [reply]
@yousefelgoharyx have you tried taking a look at this thread? → https://github.com/diegomura/react-pdf/issues/2350

**Reply by yousefelgoharyx** [reply]
I did my research before submitting the issue
the thread you're referring to isn't related to my use case 
To simplify things
I updated the codesandbox
Please take a look at /app/api/render/route.tsx
Try rendering `<PDFWithoutContext />` and try rendering `<PDF />`
You will see that `<PDFWithoutContext />` works just fine unlike `<PDF />` which throws an error
So the issue as I described above is that there's a problem when using context in rendering the pdf
not a problem in using context in the app router
all im saying is that the problem is caused by app router

**Reply by yousefelgoharyx** [reply]
If I managed to convince you that there's an issue
maybe we can discuss where I think the problem is

**Reply by yousefelgoharyx** [reply]
One more thing to prove my point
this is the error message
`Error: Minified React error #130; visit https://reactjs.org/docs/error-decoder.html?invariant=130&args[]=object&args[]= for the full message or use the non-minified dev environment for full errors and additional helpful warnings.`

when you visit that react page you will see that the original error is
`Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: object.[missing argument]`

that's why I think next is missing with every import in the project (replacing client components with a reference I believe)
and to further prove my point
this whole scenario works just fine in the pages directory
and I assume that's because next is not missing with any imports there

**Reply by samcx** [reply]
@yousefelgoharyx Switched it up slightly in my testing.

```
import { NextResponse } from "next/server";
import { renderToBuffer } from "@react-pdf/renderer";
// import { PDF } from "../../../components/PDF";
import { PDFWithoutContext } from "@/components/PDFWithoutContext";

export const GET = async () => {
  const pdf = await renderToBuffer(<PDFWithoutContext />);

  const headers = new Headers();
  headers.set('Content-Type', 'application/pdf');
  headers.set('Content-Disposition', 'attachment; filename="document.pdf"');

  return new NextResponse(pdf, { headers });
};
```

```
"use client";

import { useRouter } from 'next/navigation'

export default function Home() {
  const router = useRouter();
  
  return (
    <main className="flex min-h-screen flex-col items-center justify-between p-24">
      <button
        onClick={() => {
          router.push('/api/render');
        }}
      >
        Create
      </button>
    </main>
  );
}
```

This works (the browser automatically downloads the pdf as an attachment), and as expected, using `<PDF />` doesn't work. It shows this error. I'm not getting the Minified React error.

```
Error: Attempted to call PDF() from the server but PDF is on the client. It's not possible to invoke a client function from the server, it can only be rendered as a Component or passed to props of a Client Component.
```

This is expected with App Router—you can't pass a Client Component to a Route Handler (operating on the server, outside the scope of your React tree), which is not the same as passing a component (not using React Server Components) to an API Route in Pages Router.

Regardless of whether you used App Router or Pages Router—I'm confused about is why you want to use Context in `<PDF />`—wouldn't you expect that component to be a component separate from your React app's tree, meaning you shouldn't use Context in that case? The purpose of using Context with App Router would be to pass data from a Layout to a deeply nested Page, and likewise with Pages Router (`_app.js` to a Page).

Interestingly, the `@react-pdf/renderer` doesn't work with Next.js v15 (the latest canary) because they are using `__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;` in their internal code, which breaks with React 19.

- https://github.com/diegomura/react-pdf/issues/2756

**Reply by yousefelgoharyx** [reply]
I should be able to use context anywhere, not just with react dom
And I want to use a context with React pdf to prevent prop drilling 
If this is not a bug
Could elabo why my example would work fine in pages router?

**Reply by samcx** [reply]
> I should be able to use context anywhere

This is true, it just needs to be in the context of the React tree, creating and using it (e.g., the `<children>` need to be in the tree in order to have access to it). In your current :repro:, the `<PDF />` is not within the tree—it's a completely separate React Client Component. 

> Could elabo why my example would work fine in pages router?

Every component is either a Client or Server Component (two different types of [React Server Components](https://react.dev/reference/rsc/server-components)) in App Router. In Pages Router, there is no concept of RSC. That's why your `<PDF />` component, _which just utilizes React Context within the component file itself_, works.

The italics is the problem—you have to import a separate component to your Route Handler or API Route with React PDF. That separate component lives outside of your React tree, so it wouldn't be possible to pass anything from your Layout or `_app.js`. You would have to get creative (e.g., [createElement](https://react.dev/reference/react/createElement)) within your Route Handler or API Route.
